bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitTemperatureSensorImpl.java

```diff
@@ -58,17 +60,23 @@ public void unsubscribeCurrentTemperature() {
 
     protected double convertToCelsius(double degrees) {
         if (getSettings().useFahrenheitTemperature) {
-            return Math.round((5d / 9d) * (degrees - 32d) * 1000d) / 1000d;
+            double rawConvert = (5d / 9d) * (degrees - 32d);
+            BigDecimal roundConvert = new BigDecimal(Double.toString(rawConvert)).setScale(1, RoundingMode.HALF_UP);
+            return roundConvert.doubleValue();
         } else {
-            return degrees;
+            BigDecimal roundConvert = new BigDecimal(Double.toString(degrees)).setScale(1, RoundingMode.HALF_UP);
+            return roundConvert.doubleValue();
         }
```

Comment:
It bothers me that you have these duplicate methods for converting to and from Celsius.
For the sake of refactoring I'd prefer if you use this utility method to handle all your conversions and rounding.
```

    private <T extends Quantity<T>> double convertAndRound(double value, Unit<T> from, Unit<T> to) {
        double rawValue = from == to ? value : from.getConverterTo(to).convert(value);
        return new BigDecimal(rawValue).setScale(1, RoundingMode.HALF_UP).doubleValue();
    }
```
Then you can make your methods look like this:
```
protected double convertToCelsius(double degrees){
     return convertAndRound(degrees, getSettings().useFahrenheitTemperature ? ImperialUnits.FAHRENHEIT : SIUnits.CELSIUS, SIUnits.CELSIUS);
}

protected double convertFromCelsius(double degrees){
     return convertAndRound(degrees, getSettings().useFahrenheitTemperature ? SIUnits.CELSIUS : ImperialUnits.FAHRENHEIT, ImperialUnits.FAHRENHEIT);
}
```

Also, I would like you to find someway to get rid of the code duplication of these methods in `HomekitTemperatureSensorImpl` and `HomekitTermostatImpl`. Either put these methods in a shared superclass or put them in a utility class.

Reply:
Is that better?